### ISSUE TITLE
HTTP Error 422 when key is not set

### ISSUE THREAD

**Issue by dwene** [aaimio/get-persistent-value#2]
Hi Aaron,

Nice job with this tool, thanks for making it available to us! For my use case, I'm trying to see if I need to perform an action I can only do once per PR. 

I was trying to generate a key based on the PR branch, and use this action to attempt to fetch the value. If the value isn't there, I know I need to perform the action. But when I try to use it this way, my action fails with HTTP status code 422, which is an odd one. I would expect a 404 if a key doesn't exist instead of 422. Either way, I find it strange that the job fails if a key doesn't exist, instead of returning null or empty string or something, and makes my use case impossible with this action.

I can think of a couple of solutions:
1. Add another input that allows silent fails, which just return empty string.
2. Change the API to return 404 when a key doesnt exist, and tweak the action to return null/empty string when a 404 is received.

I was thinking about forking the API to make this fix myself but noticed the API isn't open sourced. Any chance you'd be interested in open sourcing it?

Thanks!

**Comment by aaimio** [comment]
Hi @dwene, thanks for opening an issue, I appreciate it!

Would you be able to share an example workflow for my reference? I would like to understand how this part works:

> I was trying to generate a key based on the PR branch

- On the API side, I'll update the HTTP status code for when a key couldn't be found
- On the action side, I'll add an extra input that allows silent failures, it would return an empty string or perhaps a default value?

Definitely interested in open sourcing the API but I haven't gotten to it yet.

Hope to hear from you soon.

**Comment by dwene** [comment]
Hey @aaimio,

Thanks for the quick response! Sure here's my latest action I'm building. Hopefully it makes sense. 

I'm using Helm to deploy some docker containers on our kubernetes cluster for every PR, as a QA environment. I need to run `Helm Install` the first time, and `Helm Update` every other time a commit is pushed. 

I was able to workaround the issue I was describing above by using `curl` and adding some logic to handle failures.

```
        DEPLOYED_VALUE=`curl -X GET \
        -H 'x-api-key: ${{ inputs.PERSISTENT_VALUE_ACCESS_TOKEN }}' \
        -H 'x-github-repo: lambdacurry/nibll-home' \ 
        'https://persistent.aaim.io/api/values/get?key=charts-nibll-home-${{ inputs.DEPLOY_ALIAS }}-api&output=plain' || echo 'not_deployed'`
        echo '::set-output name=value::$DEPLOYED_VALUE'
```

Here's  the full composite action. 

```
name: 'Deploy Home API'
description: 'Deploy Home API'

inputs:
  GITHUB_TOKEN:
    description: 'github token needed for docker credentials'
    required: true

  LAMBDACURRY_CHART_PASSWORD:
    description: 'password for lambdacurry chart'
    required: true

  KUBECONFIG:
    description: 'kubeconfig file'
    required: true

  POSTGRES_PASSWORD:
    description: 'postgres password'
    required: true

  REDIS_PASSWORD:
    description: 'redis password'
    required: true

  MAIL_USER:
    description: 'mail user'
    required: true

  MAIL_PASS:
    description: 'mail password'
    required: true

  OAUTH_FACEBOOK_CLIENT_SECRET:
    description: 'facebook client secret'
    required: true

  OAUTH_GOOGLE_CLIENT_SECRET:
    description: 'google client secret'
    required: true

  PERSISTENT_VALUE_ACCESS_TOKEN:
    description: 'persistent value access token'
    required: true

  DEPLOY_ALIAS:
    description: 'deploy alias'
    required: true

runs:
  using: 'composite'

  steps:
    - name: Get Deployed Chart
      id: chart_deployed
      shell: 'bash'
      run: |-
        DEPLOYED_VALUE=`curl -X GET \
        -H 'x-api-key: ${{ inputs.PERSISTENT_VALUE_ACCESS_TOKEN }}' \
        -H 'x-github-repo: lambdacurry/nibll-home' \ 
        'https://persistent.aaim.io/api/values/get?key=charts-nibll-home-${{ inputs.DEPLOY_ALIAS }}-api&output=plain' || echo 'not_deployed'`
        echo '::set-output name=value::$DEPLOYED_VALUE'

    - name: Generate Helm Values File
      shell: 'bash'
      run: 'yarn home-api-generate-helm-yaml'
      env:
        POSTGRES_PASSWORD: ${{ inputs.POSTGRES_PASSWORD }}
        REDIS_PASSWORD: ${{ inputs.REDIS_PASSWORD }}
        MAIL_USER: ${{ inputs.MAIL_USER }}
        MAIL_PASS: ${{ inputs.MAIL_PASS }}
        OAUTH_FACEBOOK_CLIENT_SECRET: ${{ inputs.OAUTH_FACEBOOK_CLIENT_SECRET }}
        OAUTH_GOOGLE_CLIENT_SECRET: ${{ inputs.OAUTH_GOOGLE_CLIENT_SECRET }}
        GIT_COMMIT: ${{ github.sha }}
        DEPLOY_ALIAS: ${{ inputs.DEPLOY_ALIAS }}

    - name: 'Build Home API'
      shell: 'bash'
      run: 'npx nx run home-api:build'

    - uses: whoan/docker-build-with-cache-action@v5
      with:
        username: '${{ github.actor }}'
        password: '${{ inputs.GITHUB_TOKEN }}'
        image_name: lambda-curry/nibll-home-api
        image_tag: ${{ github.sha }},${{ inputs.DEPLOY_ALIAS }}
        push_git_tag: true
        registry: ghcr.io
        context: .
        dockerfile: apps/home/api/Dockerfile

    - name: Helm Install
      uses: WyriHaximus/github-action-helm3@v2
      if: ${{ steps.chart_deployed.outputs.value != 'deployed' }}
      with:
        exec: |
          helm repo add lambdacurry https://charts.lambdacurry.dev --username lambdacurry --password ${{ inputs.LAMBDACURRY_CHART_PASSWORD }} && \
          helm install ${{ inputs.DEPLOY_ALIAS }}-api \
          --atomic \
          --create-namespace \
          --version=0.0.2 \
          -n nh-${{ inputs.DEPLOY_ALIAS }} \
          -f api-values.yaml \
          lambdacurry/nibll-home
        kubeconfig: '${{ inputs.KUBECONFIG }}'

    - name: Helm Update
      uses: WyriHaximus/github-action-helm3@v2
      if: ${{ steps.chart_deployed.outputs.value == 'deployed' }}
      with:
        exec: |
          helm repo add lambdacurry https://charts.lambdacurry.dev --username lambdacurry --password ${{ inputs.LAMBDACURRY_CHART_PASSWORD }} && \
          helm upgrade ${{ inputs.DEPLOY_ALIAS }}-api \
          --cleanup-on-fail \
          --version=0.0.2 \
          -n nh-${{ inputs.DEPLOY_ALIAS }} \
          -f api-values.yaml \
          lambdacurry/nibll-home
        kubeconfig: '${{ inputs.KUBECONFIG }}'

    - name: Mark Chart Deployed
      uses: aaimio/set-persistent-value@v1.1.0
      with:
        key: charts-nibll-home-${{ inputs.DEPLOY_ALIAS }}-api
        value: deployed
        access_token: ${{ inputs.PERSISTENT_VALUE_ACCESS_TOKEN }}
```

**Comment by aaimio** [comment]
> Hey @aaimio,
> 
> Thanks for the quick response! Sure here's my latest action I'm building. Hopefully it makes sense.
> 
> I'm using Helm to deploy some docker containers on our kubernetes cluster for every PR, as a QA environment. I need to run `Helm Install` the first time, and `Helm Update` every other time a commit is pushed.
> 
> I was able to workaround the issue I was describing above by using `curl` and adding some logic to handle failures.

Glad to hear you were able to find a quick fix! And thank you for sending through the entire workflow.

I've made some changes to the API (changes are already deployed) and have [released a new version (v1.1.1)](https://github.com/aaimio/get-persistent-value/releases/tag/v1.1.1), that should address the issue you were facing. I've also [bumped `set-persistent-value` action to `v1.1.1`](https://github.com/aaimio/set-persistent-value/releases/tag/v1.1.1) to stay in sync with the `get-persistent-value` version. 

Let me know if this works (or doesn't) and I'll get back to you asap.

Thanks again!
